pkg/tokenreviews/apiserver.go

```diff
@@ -0,0 +1,187 @@
+		GenericAPIServer: genericServer,
+	}
+
+	v1Storage, postStartHooks, err := c.V1RESTStorage()
+	if err != nil {
+		return nil, err
+	}
+
+	apiGroupInfo := genericapiserver.NewDefaultAPIGroupInfo(kauthv1.GroupName, serverscheme.Scheme, metav1.ParameterCodec, serverscheme.Codecs)
+	apiGroupInfo.VersionedResourcesStorageMap[kauthv1.SchemeGroupVersion.Version] = v1Storage
+	if err := s.GenericAPIServer.InstallAPIGroup(&apiGroupInfo); err != nil {
+		return nil, err
+	}
+
+	for hookname := range postStartHooks {
+		s.GenericAPIServer.AddPostStartHookOrDie(hookname, c.ExtraConfig.postStartHooks[hookname])
+	}
+
+	return s, nil
+}
+
+func (c *completedConfig) V1RESTStorage() (map[string]rest.Storage, map[string]genericapiserver.PostStartHookFunc, error) {
+	c.ExtraConfig.makeV1Storage.Do(func() {
+		c.ExtraConfig.v1Storage, c.ExtraConfig.postStartHooks, c.ExtraConfig.v1StorageErr = c.newV1RESTStorage()
+	})
+
+	return c.ExtraConfig.v1Storage, c.ExtraConfig.postStartHooks, c.ExtraConfig.v1StorageErr
+}
+
+func (c *completedConfig) tokenReviewStorage(
+	corev1Client corev1.CoreV1Interface,
+	oauthClient *oauthclients.Clientset,
+	userClient *userclients.Clientset,
+) (rest.Storage, map[string]genericapiserver.PostStartHookFunc, error) {
+	bootstrapUserDataGetter := bootstrap.NewBootstrapUserDataGetter(corev1Client, corev1Client)
+
+	// create informer for the users to be used in user <-> groups mapping
+	userInformer := userinformer.NewSharedInformerFactory(userClient, defaultInformerResyncPeriod)
+	if err := userInformer.User().V1().Groups().Informer().AddIndexers(cache.Indexers{
+		usercache.ByUserIndexName: usercache.ByUserIndexKeys,
+	}); err != nil {
+		return nil, nil, err
+	}
+
+	groupMapper := usercache.NewGroupCache(userInformer.User().V1().Groups())
+	oauthInformer := oauthinformer.NewSharedInformerFactory(oauthClient, defaultInformerResyncPeriod)
+
+	timeoutValidator := tokenvalidators.NewTimeoutValidator(
+		oauthClient.OauthV1().OAuthAccessTokens(),
+		oauthInformer.Oauth().V1().OAuthClients().Lister(),
+		c.ExtraConfig.AccessTokenInactivityTimeout,
+		minimumInactivityTimeoutSeconds)
+
+	if c.ExtraConfig.postStartHooks == nil {
+		c.ExtraConfig.postStartHooks = map[string]genericapiserver.PostStartHookFunc{}
+	}
+
+	c.ExtraConfig.postStartHooks["openshift.io-StartUserInformer"] = func(ctx genericapiserver.PostStartHookContext) error {
+		go userInformer.Start(ctx.StopCh)
+		return nil
+	}
+	c.ExtraConfig.postStartHooks["openshift.io-StartOAuthInformer"] = func(ctx genericapiserver.PostStartHookContext) error {
+		go oauthInformer.Start(ctx.StopCh)
+		return nil
+	}
+	c.ExtraConfig.postStartHooks["openshift.io-StartTokenTimeoutUpdater"] = func(ctx genericapiserver.PostStartHookContext) error {
+		go timeoutValidator.Run(ctx.StopCh)
+		return nil
+	}
+
+	tokenAuthenticator := tokenvalidation.NewTokenValidationHandler(
+		oauthClient.OauthV1().OAuthAccessTokens(), bootstrapUserDataGetter, userClient.UserV1().Users(), groupMapper,
```

Comment:
is this congruent to what we have in o/k today?  I'm surprised we use a real client to get oauth tokens.

Reply:
yes, apparently we're still doing live lookups on authentication request:
https://github.com/openshift/kubernetes/blob/4b40bb4792d0b183fd02636660c55faac08a8735/pkg/kubeapiserver/authenticator/patch_authenticator.go#L79